src/transformStateWithClones.js

```diff
@@ -7,7 +7,59 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const statesHistory = [];
+  let stateCopy = { ...state };
+
+  for (const action of actions) {
+    stateCopy = { ...stateCopy };
+
+    switch (action.type) {
+      case 'addProperties':
+        stateCopy = addProperties(stateCopy, action.extraData);
+        break;
+
+      case 'removeProperties':
+        stateCopy = removeProperties(stateCopy, action.keysToRemove);
+        break;
+
+      case 'clear':
+        stateCopy = {};
```

Comment:
[DONT REPEAT YOURSELF] - You're assigning the result of the operation to `stateCopy` in every case of the switch. You can perform this assignment just once after the switch statement, with the result of the operation.